src/database.h

```diff
@@ -37,16 +38,46 @@ class Database
 	virtual void beginSave() {}
 	virtual void endSave() {}
 
-	virtual bool saveBlock(const v3s16 &pos, const std::string &data) = 0;
-	virtual void loadBlock(const v3s16 &pos, std::string *block) = 0;
-	virtual bool deleteBlock(const v3s16 &pos) = 0;
+	bool saveBlock(const v3s16 &pos, const std::string &data)
+	{
+		m_blocks_not_found.erase(pos);
+		return dbSaveBlock(pos, data);
+	}
+	void loadBlock(const v3s16 &pos, std::string *block)
+	{
+		if (m_blocks_not_found.count(pos)) {
+			block->clear();
+		} else {
+			dbLoadBlock(pos, block);
+			if (!block->length())
+				m_blocks_not_found.insert(pos);
+		}
+	}
+	bool deleteBlock(const v3s16 &pos)
+	{
+		bool deleted = dbDeleteBlock(pos);
+		if (deleted)
+			m_blocks_not_found.insert(pos);
+		return deleted;
+	}
+	bool blockNotFound(const v3s16 &pos) const
+		{ return m_blocks_not_found.count(pos); }
+	void forgetBlocksNotFound() { m_blocks_not_found.clear(); }
 
 	static s64 getBlockAsInteger(const v3s16 &pos);
 	static v3s16 getIntegerAsBlock(s64 i);
 
 	virtual void listAllLoadableBlocks(std::vector<v3s16> &dst) = 0;
 
 	virtual bool initialized() const { return true; }
+
+protected:
+	virtual bool dbSaveBlock(const v3s16 &pos, const std::string &data) = 0;
+	virtual void dbLoadBlock(const v3s16 &pos, std::string *block) = 0;
+	virtual bool dbDeleteBlock(const v3s16 &pos) = 0;
```

Comment:
I don't like these names.  `loadBlockFromDatabase` and the like would be better.